### PR TITLE
fix: use API field names in structured output schemas

### PR DESCRIPTION
**Pull request by jbeckwith-oai**
## Summary

- Generate structured-output JSON Schema properties, required names, and reused-definition paths from each field's canonical `api_name`.
- Reject duplicate API aliases explicitly instead of silently collapsing schema properties.
- Encode each definition path segment separately to prevent direct-alias/nested-path collisions, escape complete `$defs` names as RFC 6901 JSON Pointer tokens, and percent-encode URI fragments correctly.
- Support aliases containing `/`, `~`, `#`, spaces, `%`, reserved URI characters, and Unicode while preserving valid `$defs` references.
- Keep internal `api_name` metadata out of emitted schemas while preserving descriptions, nullable fields, Ruby readers, API parsing, and request serialization.
- Remove the redundant converter dispatch branch and reuse the existing canonical JSON Schema converter.
- Add isolated regression coverage for aliases, duplicate aliases, pointer/URI escaping, deep nested/reused definitions, colliding direct/nested paths, Ruby/API-name collisions, converter round trips, and real WebMock-backed Chat Completions and Responses request/response round trips.

## Castiron / generation ownership

**No upstream Castiron generator, compiler, renderer, template, OpenAPI schema, or companion regeneration change is required.**

Verified against the actual Castiron generation workflow, not just file appearance:

1. `docs/stainless-sdk-json-fixtures/release-primary/generated-file-excludes.yaml` explicitly excludes Ruby `lib/openai/helpers/` and broad `test/`, while explicitly re-including only generated resource tests and namespace scaffolding.
2. Castiron's `crates/castiron-verify/src/generated_files.rs` implements the exact prefix/include-prefix precedence used by generated-tree verification.
3. Evaluating that real policy against every changed path and a generated control path gives:

```
EXCLUDED lib/openai/helpers/structured_output/base_model.rb
EXCLUDED lib/openai/helpers/structured_output/json_schema_converter.rb
EXCLUDED test/openai/helpers/structured_output_api_names_test.rb
GENERATED test/openai/resources/chat/completions_test.rb
```

4. Castiron's Ruby renderer (`crates/castiron-render-ruby/src/output.rs` and `src/lib/render_views.rs`) emits package/client/model/resource/signature files and resource tests, not structured-output helper code or handwritten helper tests.
5. The SDK's `CONTRIBUTING.md` independently states that the generator never modifies `lib/openai/helpers/`.

The overlapping #385 changes unrelated Tapioca/reader behavior; this PR does not incorporate or modify any of that work.

## Regression evidence

The initial suite failed against unmodified `origin/main` with **5 failures in 6 tests**: invalid property names, leaked `api_name`, lost nested `$defs`, and incorrect Chat Completions/Responses request schemas. Subsequent reviewer regressions independently failed before their fixes for duplicate aliases, JSON Pointer escaping, URI-fragment escaping, deeply nested definitions, and direct-alias/nested-path collisions.

## Verification

- `mise exec ruby@4.0.6 -- env TEST_API_BASE_URL=http://127.0.0.1:4510 ./scripts/test` — **637 runs, 2,577 assertions, 0 failures**.
- `mise exec ruby@4.0.6 -- bundle exec rake lint` — **1,390 RuboCop files clean, Sorbet clean, and 1,212 RBS files valid**.
- `mise exec ruby@4.0.6 -- bundle exec rake build:gem` — gem builds successfully.
- New alias/pointer/URI/collision regressions pass on Ruby **3.3.12**, **3.4.10**, and **4.0.6**.
- Existing structured-output suite remains green, with no new Ruby 4 URI deprecation warnings.
- Explicit thermo-nuclear code-quality review completed before every push; schema dispatch and complete reference encoding stay in their canonical owning layers.

### CODE REVIEW

File: lib/openai/helpers/structured_output/base_model.rb [hunk 1]
```diff
@@ -27,24 +27,21 @@ def to_json_schema = OpenAI::Helpers::StructuredOutput::JsonSchemaConverter.to_j
...
+                schema = OpenAI::Helpers::StructuredOutput::JsonSchemaConverter.to_json_schema_inner(
+                  type,
+                  state: new_state
+                )
                 schema = OpenAI::Helpers::StructuredOutput::JsonSchemaConverter.to_nilable(schema) if nilable
-                OpenAI::Helpers::StructuredOutput::JsonSchemaConverter.assoc_meta!(schema, meta: meta)
+                OpenAI::Helpers::StructuredOutput::JsonSchemaConverter.assoc_meta!(
+                  schema,
+                  meta: meta.except(:api_name)
+                )
```
Comment (HAYDEN-OAI) [inline lib/openai/helpers/structured_output/base_model.rb:46]: Could we reject duplicate `api_name` values instead of silently collapsing them here? For example, two Ruby fields both aliased to `:displayName` become a single schema property/required entry because of `to_h`, while model dumping likewise overwrites one value. That configuration cannot round-trip and is especially easy to introduce now that aliases are the schema keys. The collision test below covers a Ruby name matching a *different* field's API name, but not two fields sharing the same API name; a regression test plus an explicit error would keep the generated schema honest.

**Review by HAYDEN-OAI — COMMENTED** [review]
I found one additional alias-collision edge case below. The existing unresolved JSON Pointer escaping thread is also substantive: API aliases containing `/` or `~` can currently produce broken `$ref`s. I did not duplicate that inline comment.

File: lib/openai/helpers/structured_output/base_model.rb [hunk 2]
```diff
@@ -27,24 +27,21 @@ def to_json_schema = OpenAI::Helpers::StructuredOutput::JsonSchemaConverter.to_j
           def to_json_schema_inner(state:)
             OpenAI::Helpers::StructuredOutput::JsonSchemaConverter.cache_def!(state, type: self) do
               path = state.fetch(:path)
-              properties = fields.to_h do |name, field|
-                type, nilable, meta = field.fetch_values(:type, :nilable, :meta)
-                new_state = {**state, path: [*path, ".#{name}"]}
+              properties = fields.each_value.to_h do |field|
+                api_name, type, nilable, meta = field.fetch_values(:api_name, :type, :nilable, :meta)
+                new_state = {**state, path: [*path, ".#{api_name}"]}
```
[the remark this answers is not in the document]
Comment (jbeckwith-oai) [reply lib/openai/helpers/structured_output/base_model.rb:36]: Fixed in fc4fce02. The canonical JSON Schema converter now encodes the complete definition name as one RFC 6901 JSON Pointer token (`~` to `~0`, `/` to `~1`) and decodes it when materializing readable `$defs` keys. Added regressions for aliases containing both characters and for deeply nested reused definitions; the full suite passes (635 tests, 2,567 assertions).

File: lib/openai/helpers/structured_output/base_model.rb (hunk 1 above)
Comment (jbeckwith-oai) [reply lib/openai/helpers/structured_output/base_model.rb:46]: Fixed in fc4fce02. Structured-output schema generation now validates the complete field set before building properties and raises `ArgumentError` naming the duplicate API field instead of silently collapsing aliases. Added a regression where `display_name` and `legacy_name` both map to `displayName`; the existing distinct Ruby/API-name collision case remains valid, and the full suite passes (635 tests).

**Review by HAYDEN-OAI — COMMENTED** [review]
The requested duplicate-alias rejection and `/`/`~` JSON Pointer behavior are addressed, but this head still has two substantive `$ref` correctness issues already captured in the current inline threads: URI-fragment reserved characters such as `#` are not percent-encoded, and escaping only after joining path segments allows direct aliases to collide with structurally nested definition names. I did not duplicate those inline comments.

File: lib/openai/helpers/structured_output/json_schema_converter.rb [hunk 3]
```diff
@@ -98,7 +98,8 @@ def cache_def!(state, type:, &blk)
               }
               defs.store(type, stored)
               schema = blk.call
-              ref_path.replace("#/$defs/#{path.join('/')}")
+              escaped_name = path.join("/").gsub("~", "~0").gsub("/", "~1")
+              ref_path.replace("#/$defs/#{escaped_name}")
```
[the remark this answers is not in the document]
Comment (jbeckwith-oai) [reply lib/openai/helpers/structured_output/json_schema_converter.rb]: Fixed in 5657031b. After RFC 6901 token escaping, definition references are now percent-encoded as URI fragments through the supported `URI::RFC2396_PARSER`, and fragments are percent-decoded before rebuilding `$defs` keys. Added a failing-then-passing regression covering `#`, whitespace, `?`, `%`, and non-ASCII Unicode; all 637 tests and Ruby 3.3/3.4/4.0 checks pass without Ruby 4 deprecation warnings.

File: lib/openai/helpers/structured_output/json_schema_converter.rb [hunk 4]
```diff
@@ -98,7 +98,8 @@ def cache_def!(state, type:, &blk)
               }
               defs.store(type, stored)
               schema = blk.call
-              ref_path.replace("#/$defs/#{path.join('/')}")
+              escaped_name = path.join("/").gsub("~", "~0").gsub("/", "~1")
```
[the remark this answers is not in the document]
Comment (jbeckwith-oai) [reply lib/openai/helpers/structured_output/json_schema_converter.rb]: Fixed in 5657031b. Each structural path segment is escaped independently before joining, then the complete definition name is escaped again as one JSON Pointer token. This preserves boundaries: a direct alias `a/.b` now uses `$defs` key `.a~1.b`, while nested fields `a` then `b` use `.a/.b`, with distinct valid references. Added a regression that reuses both different model types and verifies both definitions plus all four references independently.

**Review by HAYDEN-OAI — APPROVED** [review]
The follow-up fixes address the remaining `$ref` correctness issues. Per-segment RFC 6901 escaping now preserves structural path boundaries before the complete `$defs` key is escaped as a single pointer token, and URI-fragment encoding safely handles reserved, percent, whitespace, and non-ASCII characters while materializing the intended readable definition key. Together with the duplicate-alias validation and Chat/Responses coverage, this looks good.
